pkg/storage/cache.go

```diff
@@ -211,28 +243,46 @@ func (c *Cache) DeleteBlob(digest, path string) error {
 			return errors.ErrCacheMiss
 		}
 
-		value := bucket.Get([]byte(path))
+		deduped := bucket.Bucket([]byte(DedupedBucket))
+		if deduped != nil {
+			if err := deduped.Delete([]byte(path)); err != nil {
+				c.log.Error().Err(err).Str("digest", digest).Str("bucket", DedupedBucket).Str("path", path).Msg("unable to delete")
 
-		if err := bucket.Delete([]byte(path)); err != nil {
-			c.log.Error().Err(err).Str("digest", digest).Str("path", path).Msg("unable to delete")
-
-			return err
+				return err
+			}
 		}
 
-		cur := bucket.Cursor()
+		origin := bucket.Bucket([]byte(OriginBucket))
+		if origin != nil {
+			originKey := c.getKey(origin)
+			if originKey != nil {
+				if err := origin.Delete([]byte(path)); err != nil {
+					c.log.Error().Err(err).Str("digest", digest).Str("bucket", OriginBucket).Str("path", path).Msg("unable to delete")
 
-		key, _ := cur.First()
-		if key == nil {
-			c.log.Debug().Str("digest", digest).Str("path", path).Msg("deleting empty bucket")
-			if err := root.DeleteBucket([]byte(digest)); err != nil {
-				c.log.Error().Err(err).Str("digest", digest).Str("path", path).Msg("unable to delete")
+					return err
+				}
 
-				return err
+				// move next candidate to origin bucket, next GetKey will return this one and storage will move the content here
```

Comment:
I understand this is for updating the boltdb content, but is there a similar logic for storage?
I mean if I have multiple `duplicate` blobs in S3 and they are empty, if the user makes a call to delete the `original` blob, do we have code copying the blob content from the `original` blob to a previously empty candidate?

Reply:
yes, we do, look here https://github.com/project-zot/zot/blob/main/pkg/storage/s3/s3.go#L1549

1. client wants to delete blob X
2. we get a blob from cache based on the digest from client
3. we delete the cache entry for 1. (here we are moving from deduped bucket to origin bucket)
4. we compare the cache entry from 2 with the blob path we received from client, if they are same then it means that is the blob with content
5. because we did 3, and 4 is true then next time we get a blobpath from cache it will be one without content. (we already moved from deduped bucket to origin bucket)
6. move content